c10/util/complex.h

```diff
@@ -0,0 +1,477 @@
+//   - There is no specialization of converting assignemnt operators, which type is
+//     convertible is soly depend on whether the scalar type is convertable
+//
+// In addition to the standard assignment, we also provide assignment operators with std and thrust
+//
+//
+// [Casting operators]
+//
+// std::complex does not have casting operators. We define casting operators casting to std::complex and thrust::complex
+//
+//
+// [Operator ""]
+//
+// std::complex has custom literals `i`, `if` and `if` defined in namespace `std::literals::complex_literals`.
+// We define our own custom literals in the namespace `c10::complex_literals`. Our custom literals does not
+// follow the same behavior as in std::complex, instead, we define _if, _id to construct float/double
+// complex literals.
+//
+//
+// [real() and imag()]
+//
+// In C++20, there are two overload of these functions, one it to return the real/imag, another is to set real/imag,
+// they are both constexpr. We follow this design.
+//
+//
+// [Operator +=,-=,*=,/=]
+//
+// Since C++20, these operators become constexpr. In our implementation, they are also constexpr.
+//
+// There are two types of such operators: operating with a real number, or operating with another complex number.
+// For the operating with a real number, the generic template form has argument type `const T &`, while the overload
+// for float/double/long double has `T`. We will follow the same type as float/double/long double in std.
+//
+// [Unary operator +-]
+//
+// Since C++20, they are constexpr. We also make them expr
+//
+// [Binary operators +-*/]
+//
+// Each operator has three versions (taking + as example):
+// - complex + complex
+// - complex + real
+// - real + complex
+//
+// [Operator ==, !=]
+// 
+// Each operator has three versions (taking == as example):
+// - complex == complex
+// - complex == real
+// - real == complex
+// 
+// Some of them are removed on C++20, but we decide to keep them
+//
+// [Operator <<, >>]
+//
+// These are implemented by casting to std::complex
+//
+//
+//
+//
+// TODO(@zasdfgbnm): c10::complex<c10::Half> is not currently supported, because:
+//  - lots of members and functions of c10::Half are not constexpr
+//  - thrust::complex only support float and double
+
+template<typename T>
+struct complex;
+
+template<typename T>
+struct alignas(sizeof(T) * 2) complex_common {
+  T storage[2];
+
+  constexpr complex_common(): storage{T(), T()} {}
+  constexpr complex_common(const T& re, const T& im = T()): storage{re, im} {}
+  template<typename U>
+  explicit constexpr complex_common(const std::complex<U> &other): complex_common(other.real(), other.imag()) {}
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  template<typename U>
+  explicit C10_HOST_DEVICE complex_common(const thrust::complex<U> &other): complex_common(other.real(), other.imag()) {}
+#endif
+
+  constexpr complex<T> &operator =(T re) {
+    storage[0] = re;
```

Comment:
You need to set the imaginary part to zero too right?

I feel like we should have tests comparing the behavior of `std::complex` and `c10::complex` on all the matching public API parts.

Reply:
Yes, you are right. I do write a test here, and if I didn't miss anything, all public APIs have test. But the test for this was poorly written: it creates a `(0, 0)` and assign `1` to it, then assert `real() == 1` and `imag() == 0`.

I have fixed this and the test for this.